International-Mentorship-ISMP/ismp
# Refactor Docker Builds and Tests

## Changes
* Add `frontend/node-modules/` to `.gitignore`
* Refactor `Makefile`
  - Rename `dev` task to `up` for greater clarity
  - Remove unnecessary `buildtest` task
  - Rename `docker` task to `docker_pull` for greater clarity
* Add `.dockerignore` to `./backend` and `./frontend` to reduce docker image size
* Refactor `backend/Dockerfile`
  - Use `debian` `slim` image to reduce docker image size and un-needed dependencies
  - Copy all source code into image
  - Remove `ENTRYPOINT`
  - Refactor `api-entrypoint.sh` into `run_tests.sh` and `run_migrations.sh`
* Refactor `docker-compose.yml` and `docker-compose.test.yml`
  - Add additional container for running database migration
* Refactor `frontend/Dockerfile`
  - Use `node:14.4.0-alpine3.12` as base image
* Ensure docker builds are idempotent
* Many more fixes & clean-ups

## Types of changes

- [x] I've labeled this PR with appropriate labels (`comp/*` component, bug/feature, documentation and other less important ones).

## Checklist
_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging your code._

- [x] I've linked the proper Github Issue(s) to my PR
- [x] I have added tests (unit, integration...etc.) that prove my fix is effective or that my change works
- [x] I have added or updated necessary documentation (if appropriate). You can search, for instance using the [`silver searcher`](https://github.com/ggreer/the_silver_searcher) `ag <STRING>` to find all usages in the repository.
- [x] I have notified the team of any breaking or important changes
